### PR TITLE
Task1

### PR DESCRIPTION
**Pull request by VasilMitovClouway**
XML And Json Implementation Ready for review

### CODE REVIEW

File: bigFile.json [hunk 1]
```diff
@@ -0,0 +1,329 @@
+{
```
Comment (freekman) [inline bigFile.json]: why is this file so big. A file like 
```js
"objects": [
 +    {
 +      "id": 1,
 +      "productName": "apple",
 +      "productPrice": 1.0
 +    },
 +    {
 +      "id": 1,
 +      "productName": "apple",
 +      "productPrice": 1.0
 +    }]
```
would be enough

File: bigFile.xml [hunk 2]
```diff
@@ -0,0 +1,328 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<products>
```
Comment (freekman) [inline bigFile.xml]: .

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  <name>MyProjectMaven</name>
+  <url>http://maven.apache.org</url>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+
+
+
+
```
Comment (freekman) [inline pom.xml]: remove the empty lines if you need to separate the dependencies by group you could use comments```<!-- -->```

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-api</artifactId>
+      <version>2.6.1</version>
+    </dependency>
+    <dependency>
```
Comment (freekman) [inline pom.xml]: just to know that we put the dependencies related to tests and the end of the pom

File: src/com/clouway/JsonCodec.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.google.gson.Gson;
+
+import java.io.*;
+
+/**
+ * @author Vasil Mitov (v.mitov.clouway@gmail.com)
+ */
+public class JsonCodec {
+  private Gson gson=new Gson();
```
Comment (freekman) [inline src/com/clouway/JsonCodec.java]: make it final

File: test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import java.nio.charset.StandardCharsets;
+
+import static org.hamcrest.CoreMatchers.is;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Vasil Mitov (v.mitov.clouway@gmail.com)
+ */
+public class JsonOperationsToMarshallAndUnmarshallObjects {
```
Comment (freekman) [inline test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java]: bad name for class that tests ```JsonCodec```

File: src/com/clouway/JsonCodec.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      writer.write(convertedObject);
+    } finally {
+      writer.close();
+    }
+    return listOfObjects.toString();
+  }
+
+
+  public ListOfObjects unmarshall(String filename) throws FileNotFoundException {
+    return gson.fromJson(new BufferedReader(new FileReader(filename+".json")),ListOfObjects.class);
```
Comment (freekman) [inline src/com/clouway/JsonCodec.java]: format

File: test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  JsonCodec jsonCodec = new JsonCodec();
+  ListOfObjects listOfObjects = new ListOfObjects();
+  FileReaderForTest fileReaderForTest = new FileReaderForTest();
+
+  /**
+   * HappyPath
+   *
```
Comment (freekman) [inline test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java]: ?

File: bigFile.json (hunk 1 above)
Comment (VasilMitovClouway) [inline bigFile.json]: It was a task to check. The time consumed to make the operations, the file size and so on on large data.

File: test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+import static org.hamcrest.CoreMatchers.is;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Vasil Mitov (v.mitov.clouway@gmail.com)
+ */
+public class JsonOperationsToMarshallAndUnmarshallObjects {
+  JsonCodec jsonCodec = new JsonCodec();
```
Comment (freekman) [inline test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java]: please make them private use the access modificators

File: test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  ListOfObjects listOfObjects = new ListOfObjects();
+  FileReaderForTest fileReaderForTest = new FileReaderForTest();
+
+  /**
+   * HappyPath
+   *
+   * @throws Exception
+   */
+  @Test
+  public void marshallingListOfObjectsToJson() throws Exception {
```
Comment (freekman) [inline test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java]: just marshal and unmarshal is good enough

File: test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java (hunk 6 above)
Comment (VasilMitovClouway) [inline test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java]: Changed

File: test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java (hunk 9 above)
Comment (VasilMitovClouway) [inline test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java]: Changed

File: test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java (hunk 10 above)
Comment (VasilMitovClouway) [inline test/com.clouway/JsonOperationsToMarshallAndUnmarshallObjects.java]: Changed

File: test/com.clouway/SpeedAndFileSizeTest.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
+package com.clouway;
+
+import org.junit.Test;
+
+/**
+ * @author Vasil Mitov (v.mitov.clouway@gmail.com)
+ */
+public class SpeedAndFileSizeTest {
```
Comment (freekman) [inline test/com.clouway/SpeedAndFileSizeTest.java]: What are you testing here speed or XmlCodec

File: src/com/clouway/XmlCodec.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+public class XmlCodec {
+  private ListOfObjects container;
+  private Object containedObjects;
+
+  public XmlCodec(ListOfObjects container, Object containedObjects) {
+    this.container = container;
+    this.containedObjects = containedObjects;
+  }
+
+  public ListOfObjects marshall(ListOfObjects objects, String fileName) throws JAXBException, FileNotFoundException {
```
Comment (freekman) [inline src/com/clouway/XmlCodec.java]: if you make it here ```arshall(Class clazz, String fileName) ``` would it not be bether

File: test/com.clouway/SpeedAndFileSizeTest.java (hunk 11 above)
Comment (VasilMitovClouway) [inline test/com.clouway/SpeedAndFileSizeTest.java]: Speed and size for XmlCodec and JsonCodec.

File: src/com/clouway/XmlCodec.java (hunk 12 above)
Comment (VasilMitovClouway) [inline src/com/clouway/XmlCodec.java]: I dont think so. Im using current contaner object.

File: bigFile.json (hunk 1 above)
Comment (freekman) [inline bigFile.json]: ok but i dont think that with this file you will be able to spot the difference but never mind leave it

File: src/com/clouway/JsonCodec.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.io.*;
+
+/**
+ * @author Vasil Mitov (v.mitov.clouway@gmail.com)
+ */
+public class JsonCodec {
+  private final Gson gson = new Gson();
+
+  public String marshall(ListOfObjects listOfObjects, String fileName) throws IOException {
```
Comment (freekman) [inline src/com/clouway/JsonCodec.java]: void

File: bigFile.json (hunk 1 above)
Comment (VasilMitovClouway) [inline bigFile.json]: I saw the difference and it was very obvious. XML File was 10 kB and Json was 3 kB.

Comment (freekman) [inline bigFile.json]: i was talking about the speed :)

File: test/com.clouway/FileReaderForTest.java [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+/**
+ * @author Vasil Mitov (v.mitov.clouway@gmail.com)
+ */
+public class FileReaderForTest {
```
Comment (freekman) [inline test/com.clouway/FileReaderForTest.java]: Just FileReader

File: test/com.clouway/FileReaderForTest.java [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  public String readFile(String path, Charset encoding)
```
Comment (freekman) [inline test/com.clouway/FileReaderForTest.java]: read

File: bigFile.json (hunk 1 above)
Comment (VasilMitovClouway) [inline bigFile.json]: XML was about 6 digits long while Json was 4 digits (in ms). While I didn't make the effort to translate it into seconds, I understood that given very large quantity of data that will grow exponentially.

**Comment by freekman** [comment]
LGTM after fix comments

File: test/com.clouway/FileReaderForTest.java (hunk 15 above)
Comment (VasilMitovClouway) [inline test/com.clouway/FileReaderForTest.java]: Changed.

File: pom.xml (hunk 4 above)
Comment (VasilMitovClouway) [inline pom.xml]: Moved to the end.

**Pull request merged**